src/reducers/profile-view.js

```diff
@@ -125,30 +130,65 @@ const symbolicationStatus: Reducer<SymbolicationStatus> = (
   }
 };
 
-const viewOptionsPerThread: Reducer<ThreadViewOptions[]> = (
-  state = [],
+export const defaultThreadViewOptions = {
+  selectedCallNodePath: [],
+  expandedCallNodePaths: new PathSet(),
+  selectedMarker: null,
+};
+
+function _getThreadViewOptions(
+  state: ThreadViewOptionsPerThreads,
+  threadsKey: ThreadsKey
+) {
+  const options = state[threadsKey];
+  if (options) {
+    return options;
+  }
+  return defaultThreadViewOptions;
+}
+
+function _updateThreadViewOptions(
+  state: ThreadViewOptionsPerThreads,
+  threadsKey: ThreadsKey,
+  updates: $Shape<ThreadViewOptions>
+): ThreadViewOptionsPerThreads {
+  const newState = { ...state };
+  newState[threadsKey] = {
+    ..._getThreadViewOptions(state, threadsKey),
+    ...updates,
+  };
+  return newState;
+}
+
+const viewOptionsPerThread: Reducer<ThreadViewOptionsPerThreads> = (
+  state = ({}: ThreadViewOptionsPerThreads),
   action
-) => {
+): ThreadViewOptionsPerThreads => {
   switch (action.type) {
     case 'PROFILE_LOADED':
-      return action.profile.threads.map(() => ({
-        selectedCallNodePath: [],
-        expandedCallNodePaths: new PathSet(),
-        selectedMarker: null,
-      }));
+      // The view options are lazily initialized. Reset to the default values.
+      return {};
     case 'BULK_SYMBOLICATION': {
       const { oldFuncToNewFuncMaps } = action;
       // For each thread, apply oldFuncToNewFuncMap to that thread's
       // selectedCallNodePath and expandedCallNodePaths.
-      return state.map((threadViewOptions, threadIndex) => {
-        const oldFuncToNewFuncMap = oldFuncToNewFuncMaps.get(threadIndex);
+      const newState = objectMap(state, (threadViewOptions, threadsKey) => {
+        // Multiple selected threads are not supported, note that transforming
+        // the threadKey with multiple threads into a number will result in a NaN.
+        // This should be fine here, as the oldFuncToNewFuncMaps only supports
+        // single thread indexes.
+        const oldFuncToNewFuncMap = oldFuncToNewFuncMaps.get(+threadsKey);
```

Comment:
I've also used `+` to parse things to number in my PR but I'm still skeptical about it. Do you think it's okay to use it or should we make the string -> integer parsing here (and in other places) more explicit with `parseInt`?

Reply:
Hmm... I think it's most likely fine. I added an assertion here to make sure we don't actually get a NaN.